### DISCUSSION TITLE
Error: ENOSPC: System limit for number of file watchers reached

### DISCUSSION THREAD

**Discussion by munchkindev** [eclipse-theia/theia#7942] [General] [answered]
### Bug Description:
The system limit for number of file watchers reached error appears when trying to host both frontend and backend on Theia using `ng serve` and `node index.js`. The backend and frontend run on ports 8080 and 4200. This is running inside of a docker container.

### Steps to Reproduce:

1. Setup your backend and frontend code on Theia.
2. Run `node index.js` on your backend
3. Run `npm install --save-dev @angular/cli` in the frontend folder and then run `./node_modules/@angular/cli/bin/ng serve --host=0.0.0.0 --disable-host-check`

### Additional Information

- Operating System: CentOS 7
- Theia Version: Theia 1.1.0

**Comment by vince-fugnitto** [comment, marked as answer]
@munchkindev the idea is to update the number of default `inotify` watchers on Linux as [documented here](https://github.com/eclipse-theia/theia/blob/bda9ff9d4cf15b28b5ffc2e984f11834adfc8f33/doc/Developing.md#linux). 

For additional information please see:
- https://github.com/theia-ide/theia-apps/pull/238#issuecomment-532925010
- https://github.com/eclipse-theia/theia/issues/5842#issuecomment-518209321

**Reply by lesolorzanov** [reply]
It is incredible that the only thing to do is to change something in your SO, instead of fixing watchpack or whatever library watches files. For some reason watchpack ignores my "ignored" options. As someone said in another comment, this is like buying a new phone because the old one is full of pictures...

**Reply by vince-fugnitto** [reply]
@zloysmiertniy I'm not sure I follow your comment.

> instead of fixing watchpack or whatever library watches files.

The framework continuously improves file-watching, and optimizes the use of handles.

For information, the framework uses [axosoft/nsfw](https://github.com/Axosoft/nsfw) for file-watching, and we have already contributed patches upstream to the repo to resolve issues we experienced. If you'd like to fix an issue please feel free to provide a pull-request upstream as well. 

> For some reason watchpack ignores my "ignored" options...

One can use the `files.watcherExclude` preference to additionally exclude other paths or files from watching.
By default the value is the following:

https://github.com/eclipse-theia/theia/blob/ab27299400b96374643b9139166afcc22195eff3/packages/filesystem/src/browser/filesystem-preferences.ts#L37-L48

Increasing the `inotify` limit is a simple procedure to increase the watcher limit which can be helpful in multiple cases, for example when dealing with a large workspace, multiple workspaces simultaneously, or different applications which consume handles. It is not a specific problem to the framework, it happens with other applications as well (vscode, intellij, ...).

**Comment by munchkindev** [comment]
@vince-fugnitto thank you for your quick response. How do I execute that properly? If I try to execute it I get an error saying that it's a read-only filesystem... Should I mendle with the docker-init somehow?

**Comment by munchkindev** [comment]
Or do I need to attach a volume with `sysctl.conf` in it?

**Comment by vince-fugnitto** [comment]
@munchkindev I believe you should update the system's default `inotify` value and not from docker, as docker likely does not expose the value to be updated (based on searching online). You are likely to find better documentation online on how to update the `inotify` value when using docker.

- https://github.com/moby/moby/issues/1044#issuecomment-273243860

**Reply by Oceanestars** [reply]
Let's say you've set it to the max (524288) and then it still gives you the watchpack error. What would be the next thing you could do?

**Reply by munchkindev** [reply]
@Oceanestars Probably take a look at what else is using up those inodes?.. Are you using Docker or running on metal?

**Reply by Oceanestars** [reply]
Both, we haven't had an issue on Docker(Yet). So it would be unusual if somehow we did max out our max_user_watches?

**Comment by munchkindev** [comment]
Setting the setting on the host system solved the issue. Thank you so much!!!
